modules/photo/src/merge.cpp

```diff
@@ -173,86 +173,94 @@ class MergeMertensImpl CV_FINAL : public MergeMertens
+
+                pow(contrast, wcon, contrast);
+                pow(saturation, wsat, saturation);
+                pow(wellexp, wexp, wellexp);
 
-            weights[i] = contrast;
-            if(channels == 3) {
-                weights[i] = weights[i].mul(saturation);
+                weights[i] = contrast;
+                if(channels == 3) {
+                    weights[i] = weights[i].mul(saturation);
+                }
+                weights[i] = weights[i].mul(wellexp) + 1e-12f;
             }
-            weights[i] = weights[i].mul(wellexp) + 1e-12f;
+        });
+        for(size_t i = 0; i < weights.size(); i++) {
             weight_sum += weights[i];
         }
         int maxlevel = static_cast<int>(logf(static_cast<float>(min(size.width, size.height))) / logf(2.0f));
         std::vector<Mat> res_pyr(maxlevel + 1);
+        std::vector<std::mutex> mutexes(maxlevel + 1);
 
-        for(size_t i = 0; i < images.size(); i++) {
-            weights[i] /= weight_sum;
-            Mat img;
-            images[i].convertTo(img, CV_32F, 1.0f/255.0f);
-
-            std::vector<Mat> img_pyr, weight_pyr;
-            buildPyramid(img, img_pyr, maxlevel);
-            buildPyramid(weights[i], weight_pyr, maxlevel);
-
-            for(int lvl = 0; lvl < maxlevel; lvl++) {
-                Mat up;
-                pyrUp(img_pyr[lvl + 1], up, img_pyr[lvl].size());
-                img_pyr[lvl] -= up;
-            }
-            for(int lvl = 0; lvl <= maxlevel; lvl++) {
-                std::vector<Mat> splitted(channels);
-                split(img_pyr[lvl], splitted);
-                for(int c = 0; c < channels; c++) {
-                    splitted[c] = splitted[c].mul(weight_pyr[lvl]);
+        parallel_for_(Range(0, static_cast<int>(images.size())), [&](const Range& range) {
+            for(int i = range.start; i < range.end; i++) {
+                weights[i] /= weight_sum;
+
+                std::vector<Mat> img_pyr, weight_pyr;
+                buildPyramid(images[i], img_pyr, maxlevel);
+                buildPyramid(weights[i], weight_pyr, maxlevel);
+
+                for(int lvl = 0; lvl < maxlevel; lvl++) {
+                    Mat up;
+                    pyrUp(img_pyr[lvl + 1], up, img_pyr[lvl].size());
+                    img_pyr[lvl] -= up;
                 }
-                merge(splitted, img_pyr[lvl]);
-                if(res_pyr[lvl].empty()) {
-                    res_pyr[lvl] = img_pyr[lvl];
-                } else {
-                    res_pyr[lvl] += img_pyr[lvl];
+                for(int lvl = 0; lvl <= maxlevel; lvl++) {
+                    std::vector<Mat> splitted(channels);
+                    split(img_pyr[lvl], splitted);
+                    for(int c = 0; c < channels; c++) {
+                        splitted[c] = splitted[c].mul(weight_pyr[lvl]);
+                    }
+                    merge(splitted, img_pyr[lvl]);
+
+                    std::lock_guard<std::mutex> guard(mutexes[lvl]);
```

Comment:
FYI: I tested this with only one mutex around a for loop and one mutex per level. Having one mutex per level is faster, as multiple threads are then allowed to write to different pyramid levels simultaneously.